### PR TITLE
fix: add v0.50 migrations; clean up schema mgmt

### LINKED ISSUE #13437: Remove `version` argument from `Service#registerSchemas()` API
### Problem

During early development of mod-service, we added the current software version as an argument to [`Service#registerSchemas()`](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-app-spi/src/main/java/com/hedera/node/app/spi/Service.java#L53). This made it easy to ensure that a service could easily register at least one guaranteed-applicable `Schema`.

But as the `SchemaRegistry` API stabilized, this was revealed as a risky and pointless addition 😓 ...in particular,
  1. Each `Schema` should be immutable once deployed, and in particular "know" the first software version it applies to.
  2. ⚠️ The class ids registered for e.g. a `KVState`'s key/value codecs **depend on the version of the `Schema` registering that `KVState`**. This means if a `Service` registers a dynamic `Schema` that _uses_ this (necessarily changing) `version` argument, the system will **lose the ability to deserialize saved states**.

### Solution

- [ ] Delete this `currentVersion` argument, ensure it is never used.
- [ ] Delete the `v0.49.0` constant [here](https://github.com/hashgraph/hedera-services/blob/develop/hedera-node/hedera-app/src/main/java/com/hedera/node/app/services/ServicesRegistryImpl.java#L45) and make it a `private static final` member of every `Schema` currently registered in the system.

### Alternatives

_No response_

### LINKED ISSUE #13538: Add migration for storage linked list repair
### Problem

Once auto-renew/expiry is enabled, we need to be able to purge all storage for a deleted contract from state. This requires the `SlotValue`s for this each contract be organized into a doubly-linked list; and for each contract's `Account` to keep a pointer to its last non-removed storage slot. Because of https://github.com/hashgraph/hedera-services/issues/13514, there are a few contracts in production where these links are not correct.

### Solution

Add a migration to analyze the slots in state and repair any broken links.

### Alternatives

_No response_

### PR DESCRIPTION
**Pull request by tinker-michaelj**
**Description**:
 - Closes #13437 by:
     1. Extracting all `Schema`s into top-level classes with internally controlled versions, deleting the `SemanticVersion` argument in `Service#registerSchemas()`. (ℹ️ This is ofc mechanical but touches many files.)
     2. Refactoring schema use analysis from `MerkleSchemaRegistry` into a separate `SchemaUseAnalysis`.
        * This refactor clarifies that a schema can be used for three purposes: `STATE_DEFINITIONS`, `MIGRATION`, and `RESTART`; and documents the conditions for each in javadoc.
 - Closes https://github.com/hashgraph/hedera-services/issues/13537
 - Closes #13538 by:
     1. Registering new `MIGRATION`-only `V0500ContractSchema` and `V0500TokenSchema` schemas to repair storage linked lists and first storage keys, respectively.
     2. Adding a `Map<String, Object> sharedValues` to the `MigrationContext` to facilitate passing information about the correct first keys from the `ContractService` to the `TokenService`.
     3. Ordering `Service` registrations by a new `Service#migrationOrder()` ordinal.

### CODE REVIEW

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/OrderedServiceMigrator.java [hunk 1]
```diff
@@ -85,7 +88,8 @@ public void doMigrations(
                 networkInfo,
                 metrics,
                 // We call with null here because we're migrating the entity ID service itself
-                null);
+                null,
+                sharedValues);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/OrderedServiceMigrator.java:92]: This `sharedValues` is only used for `SmartContract` and `TokenService` ?

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/schemas/V0490FeeSchema.java [hunk 2]
```diff
@@ -0,0 +1,109 @@
...
+    public static final String MIDNIGHT_RATES_STATE_KEY = "MIDNIGHT_RATES";
+    private static final Logger log = LogManager.getLogger(V0490FeeSchema.class);
+
+    private static ExchangeRates fs;
+
+    /**
+     * The version of the schema.
+     */
+    private static final SemanticVersion VERSION =
+            SemanticVersion.newBuilder().major(0).minor(49).patch(0).build();
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/schemas/V0490FeeSchema.java:44]: nit : we can probably add a public static field for 0_49_0_VERSION

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java [hunk 3]
```diff
@@ -149,289 +167,174 @@ public SchemaRegistry register(@NonNull Schema schema) {
...
-                () -> HapiUtils.toString(currentVersion));
-        final var latestVersion = schemasToApply.getLast().getVersion();
-        for (final var schema : schemasToApply) {
-            final var applicationType = checkApplicationType(previousVersion, latestVersion, schema);
-            logger.info("Applying {} schema {} ({})", serviceName, schema.getVersion(), applicationType);
-
+                () -> HapiUtils.toString(currentVersion),
+                () -> HapiUtils.toString(latestVersion));
+        for (final var schema : schemas) {
+            final var uses = schemaUseAnalysis.computeUses(previousVersion, latestVersion, schema);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java:217]: Probably `computeSchemaApplications` ?

**Review by lukelee-sl — DISMISSED** [review]
LGTM.  Just the smart contracts related changes.  Thank you!

File: hedera-node/hedera-file-service/src/main/java/com/hedera/node/app/service/file/FileServiceDefinition.java [hunk 4]
```diff
@@ -46,7 +49,9 @@ public final class FileServiceDefinition implements RpcServiceDefinition {
             new RpcMethodDefinition<>("systemUndelete", Transaction.class, TransactionResponse.class));
 
     private FileServiceDefinition() {
-        // Forbid instantiation
+        // Just something to keep the Gradle build believing we have a non-transitive
+        // "requires" and hence preserving our module-info.class in the compiled JAR
+        requireNonNull(CommonUtils.class);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-file-service/src/main/java/com/hedera/node/app/service/file/FileServiceDefinition.java:54]: 🤯 So weird

File: hedera-node/hedera-token-service/src/testFixtures/java/com/hedera/node/app/service/token/fixtures/FakeFeeRecordBuilder.java [hunk 5]
```diff
@@ -25,6 +28,11 @@
 public class FakeFeeRecordBuilder implements FeeRecordBuilder {
     private long transactionFee;
 
+    public FakeFeeRecordBuilder() {
+        // Just something to keep checkModuleInfo from claiming we don't require com.hedera.node.hapi
+        requireNonNull(SemanticVersion.class);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service/src/testFixtures/java/com/hedera/node/app/service/token/fixtures/FakeFeeRecordBuilder.java:33]: haha 🤯

File: hedera-node/hedera-smart-contract-service/src/main/java/com/hedera/node/app/service/contract/ContractService.java [hunk 6]
```diff
@@ -29,6 +29,20 @@
...
+     *
+     * Ensure the contract service schemas are migrated before the {@code TokenService} schemas, since
+     * the {@code TokenService} depends on the {@code ContractService} to know the updated first storage
+     * keys for contracts with broken storage links.
+     *
+     * @return {@code Integer.MIN_VALUE}
+     */
+    @Override
+    default int migrationOrder() {
+        return Integer.MIN_VALUE;
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-smart-contract-service/src/main/java/com/hedera/node/app/service/contract/ContractService.java:43]: Interesting ... 🤔 . Do you think this migrationOrder matters for any other services? Probably `FileService` should be done first before everything?

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/schemas/V0500TokenSchema.java [hunk 7]
```diff
@@ -0,0 +1,76 @@
...
+        migratedFirstKeys.forEach((contractId, firstKey) -> {
+            final var accountId = AccountID.newBuilder()
+                    .accountNum(contractId.contractNumOrThrow())
+                    .build();
+            final var account = writableAccounts.get(accountId);
+            if (account == null) {
+                log.warn("Contract account {} not found in the new state", accountId);
+            } else if (!firstKey.equals(account.firstContractStorageKey())) {
+                if (!account.smartContract()) {
+                    log.warn("Non-contract account {} has storage slots", accountId);
```
Comment (Neeharika-Sompalli) [inline hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/schemas/V0500TokenSchema.java:68]: Aren't these errors? 🤔

**Review by Neeharika-Sompalli — DISMISSED** [review]
Overall Looks great 🙌 ! I have some questions though ...

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/schemas/V0490FeeSchema.java (hunk 2 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/fees/schemas/V0490FeeSchema.java:44]: I think we should keep SPI as Hedera-agnostic as possible; so e.g. Hedera release versions don't make sense there. 🤔

File: hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java (hunk 3 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-app/src/main/java/com/hedera/node/app/state/merkle/MerkleSchemaRegistry.java:217]: Changed.

File: hedera-node/hedera-smart-contract-service/src/main/java/com/hedera/node/app/service/contract/ContractService.java (hunk 6 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-smart-contract-service/src/main/java/com/hedera/node/app/service/contract/ContractService.java:43]: Added comment to clarify.

File: hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/schemas/V0500TokenSchema.java (hunk 7 above)
Comment (tinker-michaelj) [reply hedera-node/hedera-token-service-impl/src/main/java/com/hedera/node/app/service/token/impl/schemas/V0500TokenSchema.java]: Fixed.

**Review by lukelee-sl — APPROVED** [review]
LGTM

**Pull request merged**